openfermioncirq/variational/ansatzes/low_rank.py

```diff
@@ -0,0 +1,303 @@
+                # Off-diagonal two-body coefficients for each singular vector
+                for j in range(len(self.eigenvalues)):
+                    two_body_coefficients = (
+                            self.scaled_density_density_matrices[j])
+                    if (self.include_all_cz or not numpy.isclose(
+                            two_body_coefficients[p, q], 0)):
+                        yield LetterWithSubscripts('V', p, q, j, i)
+
+    def param_bounds(self) -> Optional[Sequence[Tuple[float, float]]]:
+        """Bounds on the parameters."""
+        bounds = []
+        for param in self.params():
+            if param.letter == 'U' or param.letter == 'V':
+                bounds.append((-1.0, 1.0))
+            elif param.letter == 'T':
+                bounds.append((-2.0, 2.0))
+        return bounds
+
+    def _generate_qubits(self) -> Sequence[cirq.QubitId]:
+        """Produce qubits that can be used by the ansatz circuit."""
+        return cirq.LineQubit.range(openfermion.count_qubits(self.hamiltonian))
+
+    def operations(self, qubits: Sequence[cirq.QubitId]) -> cirq.OP_TREE:
+        """Produce the operations of the ansatz circuit."""
+
+        n_qubits = len(qubits)
+        param_set = set(self.params())
+
+        for i in range(self.iterations):
+
+            # Simulate the off-diagonal one-body terms.
+            def one_body_interaction(p, q, a, b) -> cirq.OP_TREE:
+                t_symbol = LetterWithSubscripts('T', p, q, i)
+                if t_symbol in param_set:
+                    yield XXYYGate(half_turns=t_symbol).on(a, b)
+            yield swap_network(qubits, one_body_interaction, fermionic=True)
+            qubits = qubits[::-1]
+
+            # Simulate the diagonal one-body terms.
+            for p in range(n_qubits):
+                u_symbol = LetterWithSubscripts('U', p, i)
+                if u_symbol in param_set:
+                    yield cirq.RotZGate(half_turns=u_symbol).on(qubits[p])
+
+            # Simulate each singular vector of the two-body terms.
+            prior_basis_matrix = numpy.identity(n_qubits)
+
+            for j in range(len(self.eigenvalues)):
+
+                # Get the basis change matrix.
+                basis_change_matrix = self.basis_change_matrices[j]
+
+                # Perform basis change.
+                inverse_basis_matrix = basis_change_matrix.T.conj()
+                merged_basis_matrix = numpy.dot(prior_basis_matrix,
+                                                inverse_basis_matrix)
+                yield bogoliubov_transform(qubits, merged_basis_matrix)
+
+                # Simulate the off-diagonal two-body terms.
+                def two_body_interaction(p, q, a, b) -> cirq.OP_TREE:
+                    v_symbol = LetterWithSubscripts('V', p, q, j, i)
+                    if v_symbol in param_set:
+                        yield cirq.Rot11Gate(half_turns=v_symbol).on(a, b)
```

Comment:
Again, please don't use half turns in code for quantum simulation algorithms. I can't read it. It is highly unnatural in this context.

Reply:
All gates in Cirq so far only support the use of Symbols when the angle is given in half-turns, so you'll have to accept this for now. The basic reason is that Symbols don't support arithmetic, but in the future we might come up with a better way to handle this.

Reply:
I don't know if you realize just how clunky and unnatural it is to express variational algorithms for quantum simulation in terms of exponentiated gates instead of Hamiltonian evolutions. Why would the latter need arithmetic anyways?

Reply:
I've opened https://github.com/quantumlib/Cirq/issues/868 for it, don't have time to go into detail now but we can discuss there.